kuma/core/i18n.py

```diff
@@ -1,4 +1,165 @@
+"""
+Customizations of Django i18n functions for Kuma.
+
+Django language code is lower case, like 'en-us'.
+Kuma language code is mixed case, like 'en-US'.
+"""
 from django.apps import apps
+from django.conf import settings
+from django.conf.locale import LANG_INFO
+from django.utils import lru_cache, translation
+from django.utils.translation.trans_real import (
+    check_for_language, get_languages, language_code_prefix_re,
+    language_code_re, parse_accept_lang_header)
+
+
+def django_language_code_to_kuma(lang_code):
+    """
+    Convert Django language code to Kuma language code.
+
+    Django uses lower-case codes like en-us.
+    Mozilla uses mixed-case codes like en-US.
+    """
+    return settings.LANGUAGE_URL_MAP.get(lang_code, lang_code)
+
+
+def kuma_language_code_to_django(lang_code):
+    """
+    Convert Kuma language code to Django.
+
+    Django uses lower-case codes like en-us.
+    Mozilla uses mixed-case codes like en-US.
+    """
+    return lang_code.lower()
+
+
+def get_language():
+    """Returns the currently selected language as a Kuma language code."""
+    return django_language_code_to_kuma(translation.get_language())
+
+
+@lru_cache.lru_cache(maxsize=1000)
+def get_supported_language_variant(raw_lang_code):
+    """
+    Returns the language-code that's listed in supported languages, possibly
+    selecting a more generic variant. Raises LookupError if nothing found.
+
+    If `strict` is False (the default), the function will look for an alternative
+    country-specific variant when the currently checked is not found.
+
+    lru_cache should have a maxsize to prevent from memory exhaustion attacks,
+    as the provided language codes are taken from the HTTP request. See also
+    <https://www.djangoproject.com/weblog/2007/oct/26/security-fix/>.
+
+    Based on Django 1.8.19's get_supported_language_variant from
+    django/utils/translation/trans_real.py, with changes:
+
+    * Language code can also be a Kuma language code
+    * Return Kuma languge codes
+    * Force strict=False to always allow fuzzy matching (zh-CHS gets zh-CN)
+    """
+    if raw_lang_code:
+        lang_code = kuma_language_code_to_django(raw_lang_code)
+
+        # Check for known override
+        if lang_code in settings.LOCALE_ALIASES:
+            return settings.LOCALE_ALIASES[lang_code]
+
+        # If 'fr-ca' is not supported, try special fallback or language-only 'fr'.
+        possible_lang_codes = [lang_code]
+        try:
+            possible_lang_codes.extend(LANG_INFO[lang_code]['fallback'])
+        except KeyError:
+            pass
+        generic_lang_code = lang_code.split('-')[0]
+        possible_lang_codes.append(generic_lang_code)
+        raw_supported_lang_codes = get_languages()
+        supported_lang_codes = [kuma_language_code_to_django(lang)
+                                for lang in raw_supported_lang_codes]
```

Comment:
Nit. I was thinking that it's too bad we lose the `OrderedDict` (from `get_languages()`) and its efficiency (for the following  `code in supported_lang_codes` check), when we create the `supported_lang_codes` list. What do you think of either one of the following instead?

```python
raw_supported_lang_codes = get_languages()
supported_lang_codes = OrderedDict((kuma_language_code_to_django(lang), None)
                                   for lang in raw_supported_lang_codes)
```

or, skipping the call to `get_languages` altogether:

```python
supported_lang_codes = OrderedDict((kuma_language_code_to_django(lang), None)
                                   for lang, _ in settings.LANGUAGES)
```
with both options assuming, of course, that we add a `from collections import OrderedDict`.

Reply:
I'm not sure why an ``OrderedDict`` would be more efficient to iterate over than a list.

 I'm not as sensitive to efficiency here, because of the ``lru_cache``, but it appears we could use similar code in a few places, so it makes sense to calculate it once and skip ``get_languages`` (or call it ``get_kuma_languages``).